fix(ci): retry the apt phase so a stalled mirror is survivable

#1046 이 apt 구간에 `timeout 300` 경계를 넣어 미러 정지가 job 예산 전체를 먹는 것을 막았습니다. 경계는 의도대로 작동하지만, **손해를 줄일 뿐 job 을 통과시키지는 못합니다.** 미러가 죽으면 `zsh` 설치가 실패하고 스텝이 죽는 것은 그대로입니다.

### 관측 — 경계가 발화한 실제 사례

2026-08-19, PR #1048 의 run `32223240516` 입니다. 이 브랜치의 워크플로는 머지된 main 과 바이트 동일합니다(`git diff --quiet origin/main HEAD -- .github/workflows/ci.yml` → exit 0).

```text
06:24:08.07Z  ##[group]Run sudo timeout 300 apt-get ...        ← apt 시작
06:25:18.65Z  Get:5 ... noble-security InRelease [126 kB]      ← 정지 지점
06:29:08.09Z  ##[error]Process completed with exit code 124
```

`exit code 124` 는 `timeout` 이 대상을 죽였을 때의 고유 종료 코드이고, 시작 시각 + 300초가 종료 시각과 정확히 일치합니다. job 은 15분이 아니라 5분 12초에 끝났습니다 — #1046 이 의도한 그대로입니다.

그러나 결과는 여전히 **빨간불**입니다. 스텝이 죽었으므로 뒤따르는 `npm install` 과 `scripts/run-tests.sh` 는 실행되지 않습니다. #1046 커밋 메시지의 "빨리 실패해 npm 과 스위트에 예산을 남긴다"는 논리는 예산 부분만 맞고, 남은 예산을 쓸 주체가 없다는 점이 빠져 있었습니다.

### `Acquire::Retries` 로는 안 됩니다

이미 `Acquire::Retries=3` 이 지정돼 있으나 이 정지에 걸리지 않았습니다. `apt.conf(5)` 상 이 옵션은 **다운로드 파일 단위** 재시도이고, 전송이 실패로 판정돼야 재시도가 일어납니다. 응답이 끊기지 않고 멈춰 있는 상태에서는 재시도 지점에 도달하지 못합니다 — 그래서 #1046 이 `timeout` 을 별도로 넣은 것이고, 같은 이유로 재시도도 **구간 바깥에서** 걸어야 합니다.

### 설계 후보 (확정 아님)

apt 구간 전체를 유한 재시도로 감싸되, 시도당 경계를 줄여 전체가 job 예산 안에 들어오게 합니다. 대략 이런 형태입니다.

```bash
for attempt in 1 2 3; do
  sudo timeout 120 apt-get -o Acquire::Retries=3 \
       -o Acquire::http::Timeout=20 -o Acquire::https::Timeout=20 update -y && break
  echo "apt update attempt ${attempt} failed, retrying"
  sleep 15
done
```

미해결 논점 두 가지를 PR 전에 정해야 합니다.

- 시도 횟수와 시도당 경계의 조합. 3회 × 120초 = 최대 6분이고 `test` 예산은 15분, `shellcheck` 는 10분입니다. `shellcheck` job 쪽 여유가 더 적습니다.
- 3회 모두 실패했을 때의 동작. 그대로 실패시킬지, 아니면 `zsh` 부재를 `PRAXIS_TESTS_STRICT` 가 시끄럽게 잡도록 두고 스텝은 통과시킬지 — 후자는 "조용히 건너뛰지 않는다"는 기존 설계 의도와 충돌할 수 있습니다.

외부 액션(`nick-fields/retry` 등)은 후보에서 제외했습니다. 새 의존성이 늘고, 이 리포는 액션을 SHA 로 핀하는 규약이라 유지 비용이 붙습니다.

### 적용 범위

apt 를 호출하는 job 2개 전부입니다 — `test`(예산 15분)와 `shellcheck`(예산 10분). #1046 에서 두 곳을 같이 다뤘으므로 여기서도 같이 다룹니다.

### 검증 방법

미러 정지는 원할 때 재현할 수 없으므로, 재시도 경로 자체는 실패를 강제한 조건에서 확인합니다 — 존재하지 않는 패키지나 도달 불가 미러를 지정해 재시도 횟수와 총 소요가 설계대로인지 관측합니다. 실제 미러 정지에서의 회복은 그런 사례가 다시 발생해야 확인되므로, PR 검증 코멘트의 `Unverified` 에 남깁니다.

Refs #1046

--- a/.github/workflows/ci.yml
+++ b/.github/workflows/ci.yml
@@ -53,17 +53,9 @@ jobs:
       # ubuntu-24.04 ships shellcheck, but install explicitly so the job
       # doesn't silently depend on the runner image. Rules come from
       # .shellcheckrc at the repo root.
-      # Same phase bound as the test job's apt call (#1045): this job carried the
-      # identical unguarded exposure and its own 10-minute budget to lose. The
-      # explicit install stays -- the comment above states why the job refuses to
-      # depend on the image implicitly, and that reason is unchanged.
-      - run: |
-          sudo timeout 300 apt-get -o Acquire::Retries=3 \
-                       -o Acquire::http::Timeout=20 \
-                       -o Acquire::https::Timeout=20 update -y
-          sudo timeout 300 apt-get -o Acquire::Retries=3 \
-                       -o Acquire::http::Timeout=20 \
-                       -o Acquire::https::Timeout=20 install -y shellcheck
+      # The bound from #1046 lives in scripts/ci-apt.sh now, with a retry around
+      # it (#1049): bounding alone only made this job fail faster, never pass.
+      - run: bash scripts/ci-apt.sh shellcheck
       - run: |
           find . -name '*.sh' -not -path './.git/*' -print0 \
             | xargs -0 shellcheck --severity=warning
@@ -114,22 +106,12 @@ jobs:
       # the mirror stall below. markdownlint-cli2 still comes from npm for the
       # same reason ruff is pinned: under PRAXIS_TESTS_STRICT its absence would
       # abort the run (#917).
-      # The Acquire timeouts are load-bearing (#1045). Without them apt waits
-      # indefinitely on a mirror that stops responding mid-fetch, which consumed
-      # the entire 15-minute job budget three times over two days. Retries alone
-      # do not help -- a hang never reaches the retry. Timeout covers the data
-      # transfer, not just connect, which is where the stall actually happened.
-      # Acquire::* bounds each transfer, never the phase: N stalled files still
-      # cost N * 20s, and Retries multiplies that. `timeout` bounds the phase, so
-      # a bad mirror leaves the budget for npm and the suite. sudo wraps timeout
-      # rather than the reverse -- TERM must reach apt, not a sudo that ignores it.
-      - run: |
-          sudo timeout 300 apt-get -o Acquire::Retries=3 \
-                       -o Acquire::http::Timeout=20 \
-                       -o Acquire::https::Timeout=20 update -y
-          sudo timeout 300 apt-get -o Acquire::Retries=3 \
-                       -o Acquire::http::Timeout=20 \
-                       -o Acquire::https::Timeout=20 install -y zsh
+      # The phase bound and the retry around it both live in scripts/ci-apt.sh,
+      # which carries the reasoning (#1046 bounded the stall, #1049 made it
+      # survivable). A stalled mirror was observed killing this step three times
+      # before the bound, and again after it -- bounding made the failure fast,
+      # the retry is what can make it pass.
+      - run: bash scripts/ci-apt.sh zsh
       - run: npm install -g markdownlint-cli2
       # Strict: in CI the toolchain is installed, so a SKIPPED line means the
       # run silently stopped checking something rather than accommodating a

--- a/scripts/ci-apt.sh
+++ b/scripts/ci-apt.sh
@@ -74,16 +74,18 @@ for ((attempt = 1; attempt <= attempts; attempt++)); do
     exit 0
   fi
 
-  # timeout(1) reports the two ways it can end a run with different codes, and
-  # the difference is worth naming: 124 says TERM was enough, 137 says apt
-  # ignored it and only KILL stopped it. Folding 137 into the generic branch
-  # would hide exactly the case --kill-after exists for.
+  # 124 is timeout(1)'s own code for "TERM ended the run", so it identifies a
+  # stall. 137 does not identify anything on its own -- it is the conventional
+  # 128+SIGKILL value, which the --kill-after escalation produces, but so does
+  # an OOM kill, and a command is free to exit 137 by itself. Report what was
+  # observed and list the candidates; folding it into the generic branch would
+  # still lose the one code most worth looking at.
   case "$status" in
     124)
       echo "apt attempt ${attempt} hit the ${per_attempt}s bound (exit 124 -- mirror stalled)" >&2
       ;;
     137)
-      echo "apt attempt ${attempt} ignored TERM and was killed after the ${kill_grace}s grace (exit 137)" >&2
+      echo "apt attempt ${attempt} exited 137 (SIGKILL by convention -- timeout's ${kill_grace}s escalation, an OOM kill, or apt's own status)" >&2
       ;;
     *)
       echo "apt attempt ${attempt} failed with exit ${status}" >&2

--- a/scripts/ci-apt.sh
+++ b/scripts/ci-apt.sh
@@ -18,21 +18,6 @@
 # red further down and make it harder to read.
 set -euo pipefail
 
-if [ "$#" -eq 0 ]; then
-  echo "usage: $0 <package>..." >&2
-  exit 2
-fi
-
-packages=("$@")
-
-# 3 x 120s + 2 x 15s backoff = 6m30s worst case. Measured against the two jobs
-# that call this: `test` spends ~7m on everything else against a 15-minute
-# budget, and `shellcheck` ~35s against 10 minutes, so both absorb the worst
-# case with room left. Overridable so a caller can tune without editing here.
-attempts="${CI_APT_ATTEMPTS:-3}"
-per_attempt="${CI_APT_TIMEOUT:-120}"
-backoff="${CI_APT_BACKOFF:-15}"
-
 # Per-transfer bounds, kept from #1046. They still do useful work inside an
 # attempt; they are simply not sufficient on their own.
 apt_opts=(
@@ -41,15 +26,42 @@ apt_opts=(
   -o Acquire::https::Timeout=20
 )
 
+# One attempt = update followed by install, and it must cost ONE timeout, not
+# one each: two 120s bounds in sequence make an attempt cost up to 240s, which
+# blows the per-job budget the numbers below are chosen against. The script
+# re-invokes itself in this mode so the pair runs inside a single `timeout` and
+# `apt_opts` keeps a single definition. Not a user-facing flag.
+if [ "${1:-}" = "--run-attempt" ]; then
+  shift
+  apt-get "${apt_opts[@]}" update -y
+  apt-get "${apt_opts[@]}" install -y "$@"
+  exit 0
+fi
+
+if [ "$#" -eq 0 ]; then
+  echo "usage: $0 <package>..." >&2
+  exit 2
+fi
+
+packages=("$@")
+self="$(cd "$(dirname "$0")" && pwd)/$(basename "$0")"
+
+# 3 x 120s + 2 x 15s backoff = 6m30s worst case. Measured
+# against the two jobs that call this: `test` spends ~7m on everything else
+# against a 15-minute budget, and `shellcheck` ~35s against 10 minutes, so both
+# absorb the worst case. Overridable so a caller can tune without editing here.
+attempts="${CI_APT_ATTEMPTS:-3}"
+per_attempt="${CI_APT_TIMEOUT:-120}"
+backoff="${CI_APT_BACKOFF:-15}"
+
 for ((attempt = 1; attempt <= attempts; attempt++)); do
   echo "apt attempt ${attempt}/${attempts} (timeout ${per_attempt}s): ${packages[*]}"
 
-  # `set +e` around the pair, then read the status explicitly: an `if cmd; then`
+  # `set +e` around the call, then read the status explicitly: an `if cmd; then`
   # wrapper would report the *if statement's* status (0 when the condition is
   # false and there is no else branch), which silently hides the failure.
   set +e
-  sudo timeout "$per_attempt" apt-get "${apt_opts[@]}" update -y \
-    && sudo timeout "$per_attempt" apt-get "${apt_opts[@]}" install -y "${packages[@]}"
+  sudo timeout "$per_attempt" bash "$self" --run-attempt "${packages[@]}"
   status=$?
   set -e
 

--- a/scripts/ci-apt.sh
+++ b/scripts/ci-apt.sh
@@ -1,0 +1,75 @@
+#!/usr/bin/env bash
+# Install packages with apt under a bounded, retried phase (issue #1049).
+#
+# #1046 put `timeout 300` around apt because a stalled Ubuntu mirror otherwise
+# holds the step until the job's own timeout kills it. The bound works -- it was
+# observed firing with exit 124 at exactly 300s -- but bounding alone only makes
+# the job fail faster, it never makes it pass. This script adds the missing half:
+# a finite retry around the whole phase.
+#
+# Acquire::Retries cannot do this. It retries a failed *download*, so it needs a
+# transfer to be judged failed; a mirror that accepts the connection and then
+# stops responding never reaches that point. The stall has to be cut from
+# outside, and so does the retry.
+#
+# All attempts failing is a hard failure on purpose. If apt genuinely cannot
+# install the package, run-tests.sh under PRAXIS_TESTS_STRICT aborts on the
+# missing tool anyway (#917), so passing the step here would only move the same
+# red further down and make it harder to read.
+set -euo pipefail
+
+if [ "$#" -eq 0 ]; then
+  echo "usage: $0 <package>..." >&2
+  exit 2
+fi
+
+packages=("$@")
+
+# 3 x 120s + 2 x 15s backoff = 6m30s worst case. Measured against the two jobs
+# that call this: `test` spends ~7m on everything else against a 15-minute
+# budget, and `shellcheck` ~35s against 10 minutes, so both absorb the worst
+# case with room left. Overridable so a caller can tune without editing here.
+attempts="${CI_APT_ATTEMPTS:-3}"
+per_attempt="${CI_APT_TIMEOUT:-120}"
+backoff="${CI_APT_BACKOFF:-15}"
+
+# Per-transfer bounds, kept from #1046. They still do useful work inside an
+# attempt; they are simply not sufficient on their own.
+apt_opts=(
+  -o Acquire::Retries=3
+  -o Acquire::http::Timeout=20
+  -o Acquire::https::Timeout=20
+)
+
+for ((attempt = 1; attempt <= attempts; attempt++)); do
+  echo "apt attempt ${attempt}/${attempts} (timeout ${per_attempt}s): ${packages[*]}"
+
+  # `set +e` around the pair, then read the status explicitly: an `if cmd; then`
+  # wrapper would report the *if statement's* status (0 when the condition is
+  # false and there is no else branch), which silently hides the failure.
+  set +e
+  sudo timeout "$per_attempt" apt-get "${apt_opts[@]}" update -y \
+    && sudo timeout "$per_attempt" apt-get "${apt_opts[@]}" install -y "${packages[@]}"
+  status=$?
+  set -e
+
+  if [ "$status" -eq 0 ]; then
+    echo "apt succeeded on attempt ${attempt}"
+    exit 0
+  fi
+
+  # 124 is timeout(1)'s own exit code for "I killed the child" -- worth naming,
+  # because it is the difference between a stalled mirror and a real apt error.
+  if [ "$status" -eq 124 ]; then
+    echo "apt attempt ${attempt} hit the ${per_attempt}s bound (exit 124 -- mirror stalled)" >&2
+  else
+    echo "apt attempt ${attempt} failed with exit ${status}" >&2
+  fi
+
+  if [ "$attempt" -lt "$attempts" ]; then
+    sleep "$backoff"
+  fi
+done
+
+echo "apt failed after ${attempts} attempts: ${packages[*]}" >&2
+exit 1

--- a/scripts/ci-apt.sh
+++ b/scripts/ci-apt.sh
@@ -74,13 +74,21 @@ for ((attempt = 1; attempt <= attempts; attempt++)); do
     exit 0
   fi
 
-  # 124 is timeout(1)'s own exit code for "I killed the child" -- worth naming,
-  # because it is the difference between a stalled mirror and a real apt error.
-  if [ "$status" -eq 124 ]; then
-    echo "apt attempt ${attempt} hit the ${per_attempt}s bound (exit 124 -- mirror stalled)" >&2
-  else
-    echo "apt attempt ${attempt} failed with exit ${status}" >&2
-  fi
+  # timeout(1) reports the two ways it can end a run with different codes, and
+  # the difference is worth naming: 124 says TERM was enough, 137 says apt
+  # ignored it and only KILL stopped it. Folding 137 into the generic branch
+  # would hide exactly the case --kill-after exists for.
+  case "$status" in
+    124)
+      echo "apt attempt ${attempt} hit the ${per_attempt}s bound (exit 124 -- mirror stalled)" >&2
+      ;;
+    137)
+      echo "apt attempt ${attempt} ignored TERM and was killed after the ${kill_grace}s grace (exit 137)" >&2
+      ;;
+    *)
+      echo "apt attempt ${attempt} failed with exit ${status}" >&2
+      ;;
+  esac
 
   if [ "$attempt" -lt "$attempts" ]; then
     sleep "$backoff"

--- a/scripts/ci-apt.sh
+++ b/scripts/ci-apt.sh
@@ -46,13 +46,16 @@ fi
 packages=("$@")
 self="$(cd "$(dirname "$0")" && pwd)/$(basename "$0")"
 
-# 3 x 120s + 2 x 15s backoff = 6m30s worst case. Measured
+# 3 x (120s + 10s kill grace) + 2 x 15s backoff = 7m00s worst case. Measured
 # against the two jobs that call this: `test` spends ~7m on everything else
 # against a 15-minute budget, and `shellcheck` ~35s against 10 minutes, so both
 # absorb the worst case. Overridable so a caller can tune without editing here.
 attempts="${CI_APT_ATTEMPTS:-3}"
 per_attempt="${CI_APT_TIMEOUT:-120}"
 backoff="${CI_APT_BACKOFF:-15}"
+# Without this, TERM is the only signal timeout sends: an apt that ignores or
+# delays it keeps running past the bound and overlaps the next attempt.
+kill_grace="${CI_APT_KILL_GRACE:-10}"
 
 for ((attempt = 1; attempt <= attempts; attempt++)); do
   echo "apt attempt ${attempt}/${attempts} (timeout ${per_attempt}s): ${packages[*]}"
@@ -61,7 +64,8 @@ for ((attempt = 1; attempt <= attempts; attempt++)); do
   # wrapper would report the *if statement's* status (0 when the condition is
   # false and there is no else branch), which silently hides the failure.
   set +e
-  sudo timeout "$per_attempt" bash "$self" --run-attempt "${packages[@]}"
+  sudo timeout --kill-after="$kill_grace" "$per_attempt" \
+    bash "$self" --run-attempt "${packages[@]}"
   status=$?
   set -e
 

--- a/tests/test_ci_apt.sh
+++ b/tests/test_ci_apt.sh
@@ -51,13 +51,23 @@ BIN="$STUB_ROOT/bin"
 mkdir -p "$BIN" || { echo "FAIL  [mkdir_bin]"; exit 1; }
 
 # `sudo` and `timeout` become transparent pass-throughs so the stubbed apt-get
-# is what decides each attempt's outcome. timeout drops its duration argument.
+# is what decides each attempt's outcome.
 cat > "$BIN/sudo" <<'EOF'
 #!/usr/bin/env bash
 exec "$@"
 EOF
 cat > "$BIN/timeout" <<'EOF'
 #!/usr/bin/env bash
+# Record the full invocation, then drop timeout's own arguments (any leading
+# options, then the duration) and run the command. The log is what lets a case
+# assert how many timeouts one attempt costs and whether --kill-after was set.
+echo "$*" >> "$STUB_TIMEOUT_LOG"
+while [ "$#" -gt 0 ]; do
+  case "$1" in
+    -*) shift ;;
+    *) break ;;
+  esac
+done
 shift
 exec "$@"
 EOF
@@ -81,16 +91,20 @@ if [ "$mode" = "update" ]; then
   fi
   exit "${STUB_APT_EXIT:-1}"
 fi
-# install: only reached when update succeeded
-exit 0
+# install: only reached when update succeeded. Failing here is the shape that
+# matters for the per-attempt budget -- both commands run, so an attempt that
+# bounds them separately costs two timeouts instead of one.
+exit "${STUB_APT_INSTALL_EXIT:-0}"
 EOF
 chmod +x "$BIN/sudo" "$BIN/timeout" "$BIN/apt-get"
 
 export STUB_COUNT="$STUB_ROOT/count"
+export STUB_TIMEOUT_LOG="$STUB_ROOT/timeouts"
 
 # Backoff to 0 so the suite does not spend real seconds sleeping.
 run_target() {
   : > "$STUB_COUNT"
+  : > "$STUB_TIMEOUT_LOG"
   PATH="$BIN:$PATH" \
   CI_APT_ATTEMPTS="${ATTEMPTS_OVERRIDE:-3}" \
   CI_APT_TIMEOUT=5 \
@@ -100,6 +114,7 @@ run_target() {
 }
 
 attempts_made() { cat "$STUB_COUNT" 2>/dev/null || echo 0; }
+timeouts_used() { wc -l < "$STUB_TIMEOUT_LOG" | tr -d ' '; }
 
 # 1. No packages named is a usage error, not a silent success.
 PATH="$BIN:$PATH" bash "$TARGET" >/dev/null 2>&1
@@ -115,6 +130,7 @@ run_case "exit_124_tries_3" "$(attempts_made)" "3"
 grep -q "exit 124 -- mirror stalled" "$STUB_ROOT/out"
 run_case "exit_124_names_the_stall" "$?" "0"
 
+
 # 4. A mirror that recovers mid-window is the whole point: succeed and stop.
 run_case "recovers_on_second_attempt" "$(STUB_APT_EXIT=1 STUB_APT_SUCCEED_ON=2 run_target)" "0"
 run_case "recovery_stops_early" "$(attempts_made)" "2"
@@ -126,6 +142,17 @@ run_case "clean_run_tries_once" "$(attempts_made)" "1"
 # 6. The attempt count is configurable, so the budget can be tuned per caller.
 run_case "attempts_are_configurable" "$(ATTEMPTS_OVERRIDE=2 STUB_APT_EXIT=1 run_target)" "1"
 run_case "attempts_override_respected" "$(attempts_made)" "2"
+
+# 7. One attempt costs ONE timeout, measured on the shape where it can cost two:
+#    update succeeds and install then stalls, so both commands run. Bounding
+#    them separately makes an attempt cost up to 2x per_attempt -- 3 attempts
+#    plus backoff then reach 12m30s, over the shellcheck job's 10-minute budget,
+#    and the job dies before the retry it exists to enable can finish.
+STUB_APT_SUCCEED_ON=1 STUB_APT_INSTALL_EXIT=124 run_target >/dev/null
+run_case "stalled_install_still_costs_one_timeout" "$(timeouts_used)" "3"
+run_case "stalled_install_retries_3" "$(attempts_made)" "3"
+STUB_APT_SUCCEED_ON=1 run_target >/dev/null
+run_case "clean_run_uses_one_timeout" "$(timeouts_used)" "1"
 
 echo "----"
 echo "PASS: $PASS / FAIL: $FAIL"

--- a/tests/test_ci_apt.sh
+++ b/tests/test_ci_apt.sh
@@ -154,6 +154,12 @@ run_case "stalled_install_retries_3" "$(attempts_made)" "3"
 STUB_APT_SUCCEED_ON=1 run_target >/dev/null
 run_case "clean_run_uses_one_timeout" "$(timeouts_used)" "1"
 
+# 8. TERM alone is not enough: apt that ignores it would outlive its own bound
+#    and overlap the next attempt.
+STUB_APT_EXIT=1 run_target >/dev/null
+grep -qv -- '--kill-after' "$STUB_TIMEOUT_LOG"
+run_case "kill_after_on_every_timeout" "$?" "1"
+
 echo "----"
 echo "PASS: $PASS / FAIL: $FAIL"
 if [ "$FAIL" -gt 0 ]; then

--- a/tests/test_ci_apt.sh
+++ b/tests/test_ci_apt.sh
@@ -1,0 +1,136 @@
+#!/usr/bin/env bash
+# test_ci_apt.sh — verify scripts/ci-apt.sh retries a stalled apt phase (#1049).
+#
+# The real failure cannot be reproduced on demand: it needs an Ubuntu mirror
+# that accepts the connection and then stops responding. What CAN be pinned is
+# the script's own control flow under a controlled dependency, so `sudo`,
+# `timeout` and `apt-get` are stubbed on PATH and made to fail in specific ways.
+#
+# The case that matters most is `exit_124_is_retried`: 124 is timeout(1)'s
+# "I killed the child" code and is exactly what the production stall produced
+# (observed on run 32223240516). A retry loop that treated 124 as fatal would
+# look correct in every other case and still be useless for the bug it exists
+# for.
+#
+# Usage: bash tests/test_ci_apt.sh
+# Exit:  0 = all pass; 1 = at least one fail
+
+set +e
+
+SCRIPT_DIR="$(cd "$(dirname "$0")" && pwd)"
+ROOT_DIR="$(cd "$SCRIPT_DIR/.." && pwd)"
+TARGET="$ROOT_DIR/scripts/ci-apt.sh"
+
+PASS=0
+FAIL=0
+FAILED_NAMES=()
+
+run_case() {
+  local name="$1" result="$2" expected="$3"
+  if [ "$result" = "$expected" ]; then
+    echo "PASS  [$name]"
+    PASS=$((PASS + 1))
+  else
+    echo "FAIL  [$name] expected=$expected got=$result"
+    FAIL=$((FAIL + 1))
+    FAILED_NAMES+=("$name")
+  fi
+}
+
+echo "test_ci_apt"
+
+if [ ! -f "$TARGET" ]; then
+  echo "FAIL  [target_exists] expected=yes got=no"
+  exit 1
+fi
+
+STUB_ROOT="$(mktemp -d)" || { echo "FAIL  [mktemp] could not create temp dir"; exit 1; }
+trap 'rm -rf "$STUB_ROOT"' EXIT
+
+BIN="$STUB_ROOT/bin"
+mkdir -p "$BIN" || { echo "FAIL  [mkdir_bin]"; exit 1; }
+
+# `sudo` and `timeout` become transparent pass-throughs so the stubbed apt-get
+# is what decides each attempt's outcome. timeout drops its duration argument.
+cat > "$BIN/sudo" <<'EOF'
+#!/usr/bin/env bash
+exec "$@"
+EOF
+cat > "$BIN/timeout" <<'EOF'
+#!/usr/bin/env bash
+shift
+exec "$@"
+EOF
+
+# apt-get counts its own invocations in a file and exits with whatever
+# STUB_APT_EXIT says. STUB_APT_SUCCEED_ON, when set, makes the Nth *update*
+# call (and everything after it) succeed, modelling a mirror that recovers.
+cat > "$BIN/apt-get" <<'EOF'
+#!/usr/bin/env bash
+mode=""
+for arg in "$@"; do
+  case "$arg" in
+    update|install) mode="$arg"; break ;;
+  esac
+done
+if [ "$mode" = "update" ]; then
+  n=$(( $(cat "$STUB_COUNT" 2>/dev/null || echo 0) + 1 ))
+  echo "$n" > "$STUB_COUNT"
+  if [ -n "${STUB_APT_SUCCEED_ON:-}" ] && [ "$n" -ge "$STUB_APT_SUCCEED_ON" ]; then
+    exit 0
+  fi
+  exit "${STUB_APT_EXIT:-1}"
+fi
+# install: only reached when update succeeded
+exit 0
+EOF
+chmod +x "$BIN/sudo" "$BIN/timeout" "$BIN/apt-get"
+
+export STUB_COUNT="$STUB_ROOT/count"
+
+# Backoff to 0 so the suite does not spend real seconds sleeping.
+run_target() {
+  : > "$STUB_COUNT"
+  PATH="$BIN:$PATH" \
+  CI_APT_ATTEMPTS="${ATTEMPTS_OVERRIDE:-3}" \
+  CI_APT_TIMEOUT=5 \
+  CI_APT_BACKOFF=0 \
+    bash "$TARGET" somepkg >"$STUB_ROOT/out" 2>&1
+  echo "$?"
+}
+
+attempts_made() { cat "$STUB_COUNT" 2>/dev/null || echo 0; }
+
+# 1. No packages named is a usage error, not a silent success.
+PATH="$BIN:$PATH" bash "$TARGET" >/dev/null 2>&1
+run_case "no_args_is_usage_error" "$?" "2"
+
+# 2. A mirror down for the whole window fails the step after N attempts.
+STUB_APT_EXIT=1 run_case "all_attempts_fail_exits_1" "$(STUB_APT_EXIT=1 run_target)" "1"
+run_case "all_attempts_fail_tries_3" "$(attempts_made)" "3"
+
+# 3. The production signature: timeout killed apt. Must be retried, not fatal.
+run_case "exit_124_is_retried" "$(STUB_APT_EXIT=124 run_target)" "1"
+run_case "exit_124_tries_3" "$(attempts_made)" "3"
+grep -q "exit 124 -- mirror stalled" "$STUB_ROOT/out"
+run_case "exit_124_names_the_stall" "$?" "0"
+
+# 4. A mirror that recovers mid-window is the whole point: succeed and stop.
+run_case "recovers_on_second_attempt" "$(STUB_APT_EXIT=1 STUB_APT_SUCCEED_ON=2 run_target)" "0"
+run_case "recovery_stops_early" "$(attempts_made)" "2"
+
+# 5. First-attempt success must not retry at all — the common path.
+run_case "clean_run_succeeds" "$(STUB_APT_SUCCEED_ON=1 run_target)" "0"
+run_case "clean_run_tries_once" "$(attempts_made)" "1"
+
+# 6. The attempt count is configurable, so the budget can be tuned per caller.
+run_case "attempts_are_configurable" "$(ATTEMPTS_OVERRIDE=2 STUB_APT_EXIT=1 run_target)" "1"
+run_case "attempts_override_respected" "$(attempts_made)" "2"
+
+echo "----"
+echo "PASS: $PASS / FAIL: $FAIL"
+if [ "$FAIL" -gt 0 ]; then
+  printf 'failed: %s\n' "${FAILED_NAMES[@]}"
+  exit 1
+fi
+exit 0

--- a/tests/test_ci_apt.sh
+++ b/tests/test_ci_apt.sh
@@ -130,13 +130,16 @@ run_case "exit_124_tries_3" "$(attempts_made)" "3"
 grep -q "exit 124 -- mirror stalled" "$STUB_ROOT/out"
 run_case "exit_124_names_the_stall" "$?" "0"
 
-# 3b. 137 is the other half of the same story: with --kill-after set, an apt
-#     that ignores TERM ends on KILL and reports 137, not 124. Folding it into
-#     the generic branch would hide the case --kill-after exists for.
+# 3b. 137 gets its own line because --kill-after can produce it, but the code
+#     does not prove that path: an OOM kill and a plain `exit 137` produce the
+#     same value. The message must therefore report the status and list the
+#     candidates rather than assert one.
 run_case "exit_137_is_retried" "$(STUB_APT_EXIT=137 run_target)" "1"
 run_case "exit_137_tries_3" "$(attempts_made)" "3"
-grep -q "ignored TERM and was killed" "$STUB_ROOT/out"
-run_case "exit_137_names_the_kill" "$?" "0"
+grep -q "exited 137 (SIGKILL by convention" "$STUB_ROOT/out"
+run_case "exit_137_reports_the_status" "$?" "0"
+grep -q "ignored TERM" "$STUB_ROOT/out"
+run_case "exit_137_claims_no_cause" "$?" "1"
 
 # 4. A mirror that recovers mid-window is the whole point: succeed and stop.
 run_case "recovers_on_second_attempt" "$(STUB_APT_EXIT=1 STUB_APT_SUCCEED_ON=2 run_target)" "0"

--- a/tests/test_ci_apt.sh
+++ b/tests/test_ci_apt.sh
@@ -130,6 +130,13 @@ run_case "exit_124_tries_3" "$(attempts_made)" "3"
 grep -q "exit 124 -- mirror stalled" "$STUB_ROOT/out"
 run_case "exit_124_names_the_stall" "$?" "0"
 
+# 3b. 137 is the other half of the same story: with --kill-after set, an apt
+#     that ignores TERM ends on KILL and reports 137, not 124. Folding it into
+#     the generic branch would hide the case --kill-after exists for.
+run_case "exit_137_is_retried" "$(STUB_APT_EXIT=137 run_target)" "1"
+run_case "exit_137_tries_3" "$(attempts_made)" "3"
+grep -q "ignored TERM and was killed" "$STUB_ROOT/out"
+run_case "exit_137_names_the_kill" "$?" "0"
 
 # 4. A mirror that recovers mid-window is the whole point: succeed and stop.
 run_case "recovers_on_second_attempt" "$(STUB_APT_EXIT=1 STUB_APT_SUCCEED_ON=2 run_target)" "0"

